Cruft template sync

grst opened the pull request:
Closes #35 

- [x] update bot token (needs "workflow" scope to be able to create a PR to the instance repo. 
- [x] add documentation about template sync and link it from the PR message
- [x] test (how? maybe can execute locally via act... or test it on the `infercnvpy` repo)

Zethson commented:
@grst think you can test it with workflow dispatch. Probably the safest option.

grst commented:
> @grst think you can test it with workflow dispatch. Probably the safest option.

Not really, as the action is within the cookiecutterized directory. But I did test it using `workflow_dispatch` on the [infercnvpy repo](https://github.com/icbi-lab/infercnvpy) which now uses the template, and it seems to work as expected. 

Demo PR: https://github.com/icbi-lab/infercnvpy/pull/55

grst commented:
Ok, this really seems to work as I expected. I tested the following: 

 * I migrated the `infercnvpy` repo to the cookiecutter template a while ago (some older version)
 * I copied the `sync.yaml` file to the `infercnvpy` repo
 * triggered sync via workflow dispatch
 * a pull request (https://github.com/icbi-lab/infercnvpy/pull/55) got created with the new changes since then. Merge conflicts showed up as `.rej` files. I fixed the conflicts manually and merged the PR
 * Merged #96 into the cookiecutter template - tagged as new version (v0.0.2). 
 * triggered another sync
 * a pull request (https://github.com/icbi-lab/infercnvpy/pull/56) got created with only the new changes. 

Will now go on adding some documentation, then this is ready!

giovp commented:
just understanding this better now and it's 🤯🤯🤯

grst commented on {{cookiecutter.project_name}}/docs/developer_docs.md:
```diff
@@ -240,6 +242,28 @@ in the root of the repository. Continuous integration will automatically run the
 
+### Automated template sync
+
+Automated template sync is enabled by default. This means that every night, a GitHub action runs [cruft][] to check
```
why? If there's no new release, nothing happens. 
On the other hand, if the build system is broken and fixed in the template, I would want that to be propagated asap.

Zethson replied:
Every night is completely fine. Note that "night" might not always be correct depending on your timezone :)

ivirshup reviewed (commented):
## Automatically excluded files

@grst, should we be adding some defaults for skipping here? [Cruft suggests `__init__.py` files](https://cruft.github.io/cruft/#key-features) (so code files?).

I'm wondering if we should also exclude: 

* All code files
* Content pages for docs. E.g. all the .md files except dev docs

## Release process, triggering cruft

It looks like cruft is triggered by having a release of the template made.

* Will creating a package with cookiecutter or cruft respect our git tags? Or is this just updating?
* Could there be some docs on how to make a release of the template?

ivirshup commented:
I'm not sure where the cutoff for excluded files is, so I'm not sure this is critical.

I think the versioning thing is important

grst commented:
> I'm not sure where the cutoff for excluded files is, so I'm not sure this is critical.

That's a fair point, I don't think we'd want changes to our demo `basic.py` files be propagated. I'll add these later, but I don't think it's super critical -- worst case if someone is annoyed they need to add a file manually. It's already in the dev_docs how to do that. 

> Versioning

 * If just running the `cruft create` (as indicated in the README) it uses the latest commit from the main branch (same as cookiecutter). Cruft itself does not resolve versions, but there's a `--checkout` parameter that allows to specify commits or tags. Not sure if it's an issue when the template is created from main instead of a released version. It's not a library that makes API stability guarantees after all.  
 * The automated sync automatically resolves to the tag that matches `vX.X.X` and has the highest version number. So simply merging into main does not trigger a sync. 

> Could there be some docs on how to make a release of the template?

It's literally just making a tag. Where should it go? Maybe the README? The dev docs contain documentation for an instance of the template, rather than "metadocumentation" for the template.

Zethson commented on {{cookiecutter.project_name}}/.github/workflows/sync.yaml:
```diff
@@ -0,0 +1,46 @@
+on:
+    workflow_dispatch:
+    schedule:
+        - cron: "0 2 * * *" # every night at 2:00
```
```suggestion
        - cron: "0 2 * * *" # every night at 2:00 UTC
```

Zethson commented on {{cookiecutter.project_name}}/.github/workflows/sync.yaml:
```diff
@@ -0,0 +1,46 @@
+        steps:
+            - uses: actions/checkout@v2
+            - name: Set up Python 3.10
+              uses: actions/setup-python@v2
```
```suggestion
              uses: actions/setup-python@v4
```

4.3 is the latest

Zethson commented on {{cookiecutter.project_name}}/.github/workflows/sync.yaml:
```diff
@@ -0,0 +1,46 @@
+                  branch: template-update
+                  title: Automated template update from cookiecutter-scverse
+                  body: |
+                      A new version of the [scverse cookiecutter template](https://github.com/scverse/cookiecutter-scverse)
```
Should we point to the releases of the template which hopefully also contains release notes?

Zethson reviewed (commented):
One more thing that we should consider adding to the docs: IIRC if a project has not seen activity in the last 60 days all scheduled GA wfs get disabled. This does apply to the cruft WF as well.

Excellent work @grst !

grst commented:
* Added how to create release to README
* Added `[tool.cruft]` to `pyproject.toml` to exclude certain files from syncing. 
* Updated docs as suggested by Lukas

Can I get one last approval by @Zethson and/or @ivirshup, then let's merge this! 

(I will then release `v0.0.3` and run one final sync test on infercnvpy before we release `v0.1.0` and announce it on twitter)

ivirshup commented:
@grst, please ping me for announcement when you merge and finalize

ivirshup commented:
Maybe also include #109
